src/libstd/collections/hash/raw/mod.rs

```diff
@@ -0,0 +1,1212 @@
+            }
+        }
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        // We don't have an item count, so just guess based on the range size.
+        (0, Some(unsafe { offset_from(self.end, self.ctrl) }))
+    }
+}
+
+impl<T> FusedIterator for RawIterRange<T> {}
+
+/// Iterator which returns a raw pointer to every full bucket in the table.
+pub struct RawIter<T> {
+    pub iter: RawIterRange<T>,
+    items: usize,
+}
+
+impl<T> Clone for RawIter<T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        RawIter {
+            iter: self.iter.clone(),
+            items: self.items,
+        }
+    }
+}
+
+impl<T> Iterator for RawIter<T> {
+    type Item = Bucket<T>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Bucket<T>> {
+        match self.iter.next() {
+            Some(b) => {
+                self.items -= 1;
+                Some(b)
+            }
+            None => {
+                // We don't check against items == 0 here to allow the
+                // compiler to optimize away the item count entirely if the
+                // iterator length is never queried.
+                debug_assert_eq!(self.items, 0);
+                None
+            }
+        }
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.items, Some(self.items))
+    }
+}
+
+impl<T> ExactSizeIterator for RawIter<T> {}
+impl<T> FusedIterator for RawIter<T> {}
+
+/// Iterator which consumes a table and returns elements.
+pub struct RawIntoIter<T> {
+    iter: RawIter<T>,
+    alloc: Option<(NonNull<u8>, Layout)>,
+}
+
+impl<'a, T> RawIntoIter<T> {
+    #[inline]
+    pub fn iter(&self) -> RawIter<T> {
+        self.iter.clone()
+    }
+}
+
+unsafe impl<T> Send for RawIntoIter<T> where T: Send {}
+unsafe impl<T> Sync for RawIntoIter<T> where T: Sync {}
+
+impl<T> Drop for RawIntoIter<T> {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe {
+            // Drop all remaining elements
+            if mem::needs_drop::<T>() {
+                while let Some(item) = self.iter.next() {
+                    item.drop();
+                }
+            }
+
+            // Free the table
+            if let Some((ptr, layout)) = self.alloc {
+                dealloc(ptr.as_ptr(), layout);
+            }
+        }
+    }
+}
+
+impl<T> Iterator for RawIntoIter<T> {
+    type Item = T;
+
+    #[inline]
+    fn next(&mut self) -> Option<T> {
+        unsafe { Some(self.iter.next()?.read()) }
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+}
+
+impl<T> ExactSizeIterator for RawIntoIter<T> {}
+impl<T> FusedIterator for RawIntoIter<T> {}
+
+/// Iterator which consumes elements without freeing the table storage.
+pub struct RawDrain<'a, T: 'a> {
+    iter: RawIter<T>,
+
+    // We don't use a &'a RawTable<T> because we want RawDrain to be covariant
+    // over 'a.
+    table: NonNull<RawTable<T>>,
```

Comment:
RawDrain *would* be covariant over `'a` if it contained `&'a RawTable<T>`. It would even be covariant if it was `&'a mut`. I'm guessing the former is a typo, while the latter is just a misunderstanding?

See the table and discussion here: https://doc.rust-lang.org/nightly/nomicon/subtyping.html#variance

Reply:
Right. I think that what I had in mind was covariant over `T`, not `'a`. I think the code is still correct and just the comment needs to be fixed?

Reply:
As an approximation Drain is `fn() -> T`, which can be covariant, so I suppose yes this is sound. Especially since swapping the internal pointers would be busted even if invariant.